spring-vault-core/src/main/java/org/springframework/vault/authentication/UsernamePasswordAuthentication.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.vault.authentication;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.util.Assert;
+import org.springframework.vault.VaultException;
+import org.springframework.vault.client.VaultResponses;
+import org.springframework.vault.support.VaultResponse;
+import org.springframework.vault.support.VaultToken;
+import org.springframework.web.client.HttpStatusCodeException;
+import org.springframework.web.client.RestOperations;
+
+import java.util.Collections;
+
+/**
+ * Username and password implementation of {@link ClientAuthentication}.
+ *
+ * @author Mikhael Sokolov
+ * @see UsernamePasswordAuthenticationOptions
+ * @see RestOperations
+ * @see <a href="https://www.vaultproject.io/docs/auth/userpass">Username & password</a>
+ * @since 2.4
+ */
+public class UsernamePasswordAuthentication implements ClientAuthentication {
```

Comment:
It would make sense to also add an authentication steps implementation to support reactive authentication.

Reply:
implemented for reactive as well